server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,194 @@
+         // addSubscriber call can rise exception that needs to be reported
+         // as error
+         var retStage = (v != null && !v.isEmpty())
+               ? CompletableFuture.completedFuture(v)
+               : addSubscriber(listMultimap, filterKeys, timeout, handler, ctx);
+         return retStage;
+      }), ctx, Consumers.COLLECTION_BULK_BICONSUMER);
+   }
+
+   CompletionStage<Collection<byte[]>> pollKeyValue(EmbeddedMultimapListCache<byte[], byte[]> mmList, byte[] key) {
+      return mmList.pollFirst(key, 1)
+            .thenApply((v) -> (v == null || v.isEmpty())
+                  ? null
+                  : Arrays.asList(key, v.iterator().next()));
+   }
+
+   CompletionStage<Collection<byte[]>> addSubscriber(EmbeddedMultimapListCache<byte[], byte[]> listMultimap,
+         List<byte[]> filterKeys, long timeout, Resp3Handler handler,
+         ChannelHandlerContext ctx) {
+      if (log.isTraceEnabled()) {
+         log.tracef("Subscriber for keys: " +
+               filterKeys.toString());
+      }
+      AdvancedCache<Object, Object> cache = handler.cache().withMediaType(MediaType.APPLICATION_OCTET_STREAM, null);
+      DataConversion vc = cache.getValueDataConversion();
+      PubSubListener pubSubListener = new PubSubListener(ctx.channel());
+      EventListenerKeysFilter filter = new EventListenerKeysFilter(filterKeys.toArray(byte[][]::new));
+      CacheEventConverter<Object, Object, Object> converter = (key, oldValue, oldMetadata, newValue, newMetadata,
+            eventType) -> vc.fromStorage(newValue);
+      CompletionStage<Void> addListenerStage = cache.addListenerAsync(pubSubListener, filter, converter);
+      final var pubSubFuture = pubSubListener.getFuture();
+      return CompletionStages.handleAndCompose(addListenerStage, (ignore, t) -> {
+         // If listener fails to install, complete exceptionally pubSubFuture and return
+         if (t != null) {
+            pubSubFuture.completeExceptionally(t);
+            return pubSubFuture;
+         }
+         // Listener can lose events during its install, so we need to poll again
+         // and if we get values complete the listener future. In case of exception
+         // completeExceptionally
+         return CompletionStages.handleAndCompose(pollAllKeys(listMultimap, filterKeys), (v, t2) -> {
+            // If second poll fails, remove listener and complete exceptionally
+            if (t2 != null) {
+               cache.removeListenerAsync(pubSubListener);
+               pubSubFuture.completeExceptionally(t);
+               return pubSubFuture;
+            }
+            if (v != null) {
+               // poll returns values, complete the listener future ...
+               pubSubFuture.complete(v);
+            } else {
+               // ... else wait for values with timeout if required
+               pubSubListener.startTimer(timeout);
+            }
+            pubSubFuture.thenRun(() -> {
```

Comment:
This doesn't run if the future was completed exceptionally which technically can happen inside the listener itself, which means the listener will be stuck until the timer completes (or forever if the timer didn't have a timeout)